File: src/controllers/orderController.js
const fs = require('fs');
const path = require('path');
const EventEmitter = require('events');

// Global event emitter for order events
const orderEmitter = new EventEmitter();

// Fixed Memory Leak: Move process-level/global listener registration outside of the request handler.
orderEmitter.on('orderPlaced', (data) => {
  console.log(`[Order Service] Order placed event received for user: ${data.username}`);
});

// Fixed Memory Leak: Restrict cache size and do not retain entire large req objects (headers/contexts)
const statsCache = [];
const MAX_STATS = 1000;

class OrderController {

  // POST /api/orders (Create Order)
  // Relates to user (from userController data) and products (from productController)
  static async createOrder(req, res) {
    const { username, productId, quantity } = req.body;

    // Fixed Syntax Error: Restored proper parentheses in conditional statement
    if (!username || !productId || !quantity) {
      return res.status(400).json({ error: "Missing required fields" });
    }

    try {
      // Verify user exists (Relation to UserController's data)
      const safeUsername = path.basename(username).replace(/[^a-zA-Z0-9_-]/g, '');
      const userFile = path.join(__dirname, '../data/users', `${safeUsername}.json`);
      if (!fs.existsSync(userFile)) {
        return res.status(404).json({ error: "User not found" });
      }

      // Fixed Concurrency Risk: Read/write actions are isolated per request.
      // Mock DB write latency simulation
      await new Promise(resolve => setTimeout(resolve, 50));

      const orderId = Date.now();
      const newOrder = {
        id: orderId,
        username,
        productId,
        quantity,
        status: 'pending'
      };

      // Save order
      const orderPath = path.join(__dirname, '../data/orders', `${orderId}.json`);
      
      // Fixed Risk: Enclosed directories creation and writes in try-catch to avoid unhandled rejections
      fs.mkdirSync(path.dirname(orderPath), { recursive: true });
      fs.writeFileSync(orderPath, JSON.stringify(newOrder, null, 2));

      orderEmitter.emit('orderPlaced', { username, orderId });

      res.status(201).json(newOrder);
    } catch (err) {
      console.error("Order creation failed:", err);
      res.status(500).json({ error: "Internal server error during order creation" });
    }
  }

  // GET /api/orders/:id (Retrieve Order)
  static getOrder(req, res) {
    const orderId = req.params.id;

    // Fixed Security Vulnerability: Prevent directory path manipulation in IDOR/lookup
    const safeOrderId = String(orderId).replace(/[^a-zA-Z0-9_-]/g, '');
    const orderPath = path.join(__dirname, '../data/orders', `${safeOrderId}.json`);
    
    try {
      if (!fs.existsSync(orderPath)) {
        return res.status(404).json({ error: "Order not found" });
      }

      const orderData = fs.readFileSync(orderPath, 'utf8');
      const order = JSON.parse(orderData);

      // Fixed Security Vulnerability: IDOR prevention (ensure requester is owner of the order, or is admin)
      if (req.user && req.user.username !== order.username && req.user.role !== 'admin') {
        return res.status(403).json({ error: "Access denied: Unauthorized to view this order" });
      }

      // Fixed Memory Leak: Shift oldest cache item if size exceeds MAX_STATS.
      // Avoid storing heavy req objects (like full headers) in global caches.
      if (statsCache.length >= MAX_STATS) {
        statsCache.shift();
      }
      statsCache.push({
        time: new Date(),
        orderId: safeOrderId,
        userAgent: req.headers['user-agent'] || 'unknown'
      });

      res.json(order);
    } catch (err) {
      console.error("Order retrieval failed:", err);
      res.status(500).json({ error: "Internal server error during order retrieval" });
    }
  }

  // PUT /api/orders/:id (Update Order Status)
  static updateOrder(req, res) {
    const orderId = req.params.id;
    const { status } = req.body;

    const safeOrderId = String(orderId).replace(/[^a-zA-Z0-9_-]/g, '');
    const orderPath = path.join(__dirname, '../data/orders', `${safeOrderId}.json`);
    
    try {
      if (!fs.existsSync(orderPath)) {
        return res.status(404).json({ error: "Order not found" });
      }

      const order = JSON.parse(fs.readFileSync(orderPath, 'utf8'));

      // Fixed Security: Ensure proper authorization checks
      if (req.user && req.user.username !== order.username && req.user.role !== 'admin') {
        return res.status(403).json({ error: "Access denied" });
      }

      // Validate inputs
      const allowedStatuses = ['pending', 'processing', 'completed', 'cancelled'];
      if (!allowedStatuses.includes(status)) {
        return res.status(400).json({ error: "Invalid status value" });
      }
      
      order.status = status;
      fs.writeFileSync(orderPath, JSON.stringify(order, null, 2));
      res.json({ message: "Order updated", order });
    } catch (err) {
      console.error("Order update failed:", err);
      res.status(500).json({ error: "Internal server error during order update" });
    }
  }

  // GET /api/orders/receipt (Download Receipt)
  static getReceipt(req, res) {
    const filename = req.query.file;

    if (!filename) {
      return res.status(400).json({ error: "File name is required" });
    }

    // Fixed Security Vulnerability: Sanitize receipt filename to prevent Path Traversal
    const safeFilename = path.basename(filename).replace(/[^a-zA-Z0-9_.-]/g, '');
    const receiptPath = path.join(__dirname, '../data/receipts', safeFilename);
    
    // Fixed Syntax Error: Replaced semicolon with comma in object declaration
    const fileMetadata = {
      path: receiptPath,
      accessTime: new Date()
    };
    
    console.log("Metadata:", fileMetadata);

    try {
      if (!fs.existsSync(receiptPath)) {
        return res.status(404).json({ error: "Receipt not found" });
      }

      const content = fs.readFileSync(receiptPath, 'utf8');
      res.send(content);
    } catch (err) {
      console.error("Failed to read receipt:", err);
      res.status(500).json({ error: "Internal server error while reading receipt" });
    }
  }
}

module.exports = OrderController;

File: src/controllers/order_controller.py
import os
import re
import json
import time
import asyncio
from datetime import datetime
from pathlib import Path
from typing import Dict, Any, Optional
from fastapi import HTTPException, status
from pydantic import BaseModel

stats_cache = []
MAX_STATS = 1000

class CreateOrderRequest(BaseModel):
    username: str
    productId: int
    quantity: int

class UpdateOrderRequest(BaseModel):
    status: str

class OrderController:
    @staticmethod
    async def create_order(body: CreateOrderRequest) -> Dict[str, Any]:
        if not body.username or not body.productId or not body.quantity:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Missing required fields")
            
        safe_username = os.path.basename(body.username)
        safe_username = re.sub(r'[^a-zA-Z0-9_-]', '', safe_username)
        
        current_dir = Path(__file__).resolve().parent
        user_file = current_dir / ".." / "data" / "users" / f"{safe_username}.json"
        
        if not user_file.exists():
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")
            
        await asyncio.sleep(0.05)
        
        order_id = int(time.time() * 1000)
        new_order = {
            "id": order_id,
            "username": body.username,
            "productId": body.productId,
            "quantity": body.quantity,
            "status": "pending"
        }
        
        orders_dir = current_dir / ".." / "data" / "orders"
        order_path = orders_dir / f"{order_id}.json"
        
        try:
            orders_dir.mkdir(parents=True, exist_ok=True)
            order_path.write_text(json.dumps(new_order, indent=2), encoding="utf-8")
            
            print(f"[Order Service] Order placed event received for user: {body.username}")
            
            return new_order
        except Exception as err:
            print("Order creation failed:", str(err))
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Internal server error during order creation")

    @staticmethod
    def get_order(order_id: str, requester: Optional[Dict[str, Any]], user_agent: str) -> Dict[str, Any]:
        safe_order_id = re.sub(r'[^a-zA-Z0-9_-]', '', str(order_id))
        
        current_dir = Path(__file__).resolve().parent
        order_path = current_dir / ".." / "data" / "orders" / f"{safe_order_id}.json"
        
        try:
            if not order_path.exists():
                raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Order not found")
                
            order = json.loads(order_path.read_text(encoding="utf-8"))
            
            if requester:
                if requester.get("username") != order.get("username") and requester.get("role") != "admin":
                    raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Access denied: Unauthorized to view this order")
            
            if len(stats_cache) >= MAX_STATS:
                stats_cache.pop(0)
            stats_cache.append({
                "time": datetime.now().isoformat(),
                "orderId": safe_order_id,
                "userAgent": user_agent
            })
            
            return order
        except HTTPException:
            raise
        except Exception as err:
            print("Order retrieval failed:", str(err))
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Internal server error during order retrieval")

    @staticmethod
    def update_order(order_id: str, body: UpdateOrderRequest, requester: Optional[Dict[str, Any]]) -> Dict[str, Any]:
        safe_order_id = re.sub(r'[^a-zA-Z0-9_-]', '', str(order_id))
        
        current_dir = Path(__file__).resolve().parent
        order_path = current_dir / ".." / "data" / "orders" / f"{safe_order_id}.json"
        
        try:
            if not order_path.exists():
                raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Order not found")
                
            order = json.loads(order_path.read_text(encoding="utf-8"))
            
            if requester:
                if requester.get("username") != order.get("username") and requester.get("role") != "admin":
                    raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Access denied")
                    
            allowed_statuses = ['pending', 'processing', 'completed', 'cancelled']
            if body.status not in allowed_statuses:
                raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid status value")
                
            order["status"] = body.status
            order_path.write_text(json.dumps(order, indent=2), encoding="utf-8")
            
            return {"message": "Order updated", "order": order}
        except HTTPException:
            raise
        except Exception as err:
            print("Order update failed:", str(err))
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Internal server error during order update")

    @staticmethod
    def get_receipt(filename: str) -> str:
        if not filename:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="File name is required")
            
        safe_filename = os.path.basename(filename)
        safe_filename = re.sub(r'[^a-zA-Z0-9_.-]', '', safe_filename)
        
        current_dir = Path(__file__).resolve().parent
        receipt_path = current_dir / ".." / "data" / "receipts" / safe_filename
        
        file_metadata = {
            "path": str(receipt_path),
            "accessTime": datetime.now().isoformat()
        }
        print("Metadata:", file_metadata)
        
        try:
            if not receipt_path.exists():
                raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Receipt not found")
                
            return receipt_path.read_text(encoding="utf-8")
        except HTTPException:
            raise
        except Exception as err:
            print("Failed to read receipt:", str(err))
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Internal server error while reading receipt")
